### DISCUSSION TITLE
Best way to preserve values when fields change between renders?

### DISCUSSION THREAD

**Discussion by slugmandrew** [react-hook-form/react-hook-form#6359] [Q&A] [answered]
Hi all,

😍 First time posting in this great community. Absolutely loving the library and everything about how it's managed - from the codebase to the docs, examples and tutorials it is all a _class act_, so thanks so much @bluebill1049 and all other contributors!

⚠️ Onto my problem... I am creating a dynamic form (using RHF V6), which has fields that are _conditionally rendered_ based on another field. 

**[Codesandbox here](https://codesandbox.io/s/react-hook-form-conditionally-render-svbcw)**

The form is working by rendering a list of fields, filtering them based on the dropdown field at the top. When I switch from `All` to `Subset`, only three of the fields will render (1, 2 and 4), and any field that gets re-rendered loses its value. Fields 1 and 2 are OK because they were not re-rendered.

🎥 Here's a video showing the behaviour, with and without a `key` on the fields. As you can see on field `example-4`, if the field does not have a `key`, it gets the value that used to be in `example-3`, as that now takes the same position in the DOM tree (I assume). If it does have a `key`, the value is lost entirely.

https://user-images.githubusercontent.com/153625/130788758-2cfb8109-c862-4011-9328-4f1be5b98018.mp4

**What I want to achieve is for the user to be able to enter values into all fields, switch to a different set of fields, and then switch back without losing any data.** I can think of a couple of ways to do this:

1. **Render all the fields all the time**, and just **conditionally register** them based on whether they are in the subset or not. Would this work? I'd obviously have to hide them from view and feels quite dirty.
2. **Use some kind of state to save the whole set of values**, then repopulate them when the switch happens. I'd like to avoid this if possible but I think I may have to go down this route.
3. Use **controlled components** - Pretty sure a certain person with 'blue' in their name would discourage this! 😂

If anyone's been in a similar situation any advice is greatly appreciated!

**Comment by bluebill1049** [comment, marked as answer]
hello there, thank you for the kind words and how nicely you have formatted the question. 🙏 It really helps to let others/me to help out. in V6, we unmount and remove input value by default, have you tried to set `shouldUnreigster` to `false`?

**Reply by slugmandrew** [reply]
No worries! It's actually fine taking the time to format a question well if you know you are going to get a response 🙂 Helps with thinking through the problem, too!

I was trying with `shouldUnregister: false`, but it still wouldn't preserve the values if the field was removed entirely and then added back in. Should that work? Even if a field has been totally removed from the DOM?

**Reply by bluebill1049** [reply]
yes, it should. take a look at the example:

https://codesandbox.io/s/practical-mendeleev-ccjf9?file=/src/App.js

**Reply by slugmandrew** [reply]
Awesome! So it looks like calling `filter()` on the map was the issue. I modified the CSB a little, and when I filtered the values out of the list it didn't work.

The issue I have now is that I can't `unregister` fields. I have replicated this in other situations where `shouldUnregister` is `false`. **Maybe I am misunderstanding how `unregister` works?** Please see the modified CSB: https://codesandbox.io/s/react-hook-form-unregister-v6-ts-forked-6u0lk?file=/src/index.tsx

**Reply by slugmandrew** [reply]
Hmmm... just thinking this through some more, does it even make sense for me to unregister the fields? I want to unregister it so the values aren't included in the form submission, but then that's going to cause me to lose the value inside the field, right?

Maybe the best solution is for me to do the filtering inside my submit method?

**Reply by bluebill1049** [reply]
Yes, sounds like a better approach.

**Reply by slugmandrew** [reply]
Cheers. Last question, I promise.... Just to confirm, is it expected behaviour for `unregister` to not work at all if `shouldUnregister` is `false`? This is the behaviour I am seeing in V6.

**Reply by bluebill1049** [reply]
it should work, check if your input gets register again in the render.
